### PR TITLE
client-go: add a test to document retry conditions

### PR DESCRIPTION
**Pull request by tkashem**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
add a unit test to document all the cases on which client-go retries a request.

#### Which issue(s) this PR fixes:
Fixes #

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by tkashem** [comment]
cc @wojtek-t @aojea @liggitt

File: staging/src/k8s.io/client-go/rest/with_retry.go [hunk 1]
```diff
@@ -354,6 +354,10 @@ func retryAfterResponse() *http.Response {
 }
 
 func retryAfterResponseWithDelay(delay string) *http.Response {
+	return retryAfterResponseWithCodeAndDelay(http.StatusInternalServerError, delay)
+}
+
+func retryAfterResponseWithCodeAndDelay(code int, delay string) *http.Response {
```
Comment (aojea) [inline staging/src/k8s.io/client-go/rest/with_retry.go:360]: I guess you meant to set the StatusCode based on the input var here, right?

Comment (tkashem) [reply staging/src/k8s.io/client-go/rest/with_retry.go:360]: oops, typo, fixed.

File: staging/src/k8s.io/client-go/rest/request_test.go [hunk 2]
```diff
@@ -3712,3 +3712,335 @@ func TestRequestBodyResetOrder(t *testing.T) {
...
+		{
+			name:          "GET, server returns {429, Retry-After}",
+			verb:          "GET",
+			serverReturns: responseErr{response: retryAfterResponseWithCodeAndDelay(http.StatusTooManyRequests, "0"), err: nil},
+			retryExpectation: map[string]bool{
+				"Do":     true,
+				"DoRaw":  true,
+				"Watch":  true,
+				"Stream": true,
+			},
```
Comment (roycaihw) [inline staging/src/k8s.io/client-go/rest/request_test.go:3752]:
```
map[string]bool{
				"Do":     true,
				"DoRaw":  true,
				"Watch":  true,
				"Stream": true,
			}
```
can be a reusable variable (e.g. "alwaysRetry"), similar in other places

File: staging/src/k8s.io/client-go/rest/request_test.go [hunk 3]
```diff
@@ -3712,3 +3712,335 @@ func TestRequestBodyResetOrder(t *testing.T) {
...
+			if err == nil {
+				// we need to wait here to avoid race condition.
+				<-w.ResultChan()
+			}
+		},
+	}
+
+	tests := []struct {
+		name             string
+		verb             string
```
Comment (roycaihw) [inline staging/src/k8s.io/client-go/rest/request_test.go:3738]: nit: consider turning verb into `verbs []string` to reduce the number of similar but slightly different test cases

**Review by roycaihw — COMMENTED** [review]
Maybe worth documenting in k8s.io as well? 

/lgtm
/approve

**Comment by aojea** [comment]
/lgtm
/assign @wojtek-t 
for approval

**Comment by aojea** [comment]
it would be good to have some test cases for 2xx , 3xx and 4xx responses with retry-after to check those doesn't retry

**Comment by tkashem** [comment]
> it would be good to have some test cases for 2xx , 3xx and 4xx responses with retry-after to check those doesn't retry

added these conditions to the test

File: staging/src/k8s.io/client-go/rest/request_test.go (hunk 3 above)
Comment (tkashem) [reply staging/src/k8s.io/client-go/rest/request_test.go]: done, thanks

File: staging/src/k8s.io/client-go/rest/request_test.go (hunk 2 above)
Comment (tkashem) [reply staging/src/k8s.io/client-go/rest/request_test.go]: done

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Comment by wojtek-t** [comment]
@tkashem - can you please rebase it?
This LGTM so let's have this merged :)

**Comment by tkashem** [comment]
@wojtek-t thanks for taking the time to look at it, rebase done.

**Comment by wojtek-t** [comment]
/lgtm
/approve

Thanks!

**Pull request merged**